Replace poll() with select() where poll() is not available

lithp opened the pull request.

pykello commented:
High level question: Why are we not using WaitEventSets?

I think we discussed this earlier, I just wanted to refresh my memory.

pykello commented on src/backend/distributed/connection/connection_management.c:
```diff
@@ -10,7 +10,20 @@
 
+#undef HAVE_POLL
+#undef HAVE_POLL_H
+#undef HAVE_SYS_POLL_H
```
Are these leftovers from testing?

pykello commented on src/backend/distributed/connection/connection_management.c:
```diff
@@ -10,7 +10,20 @@
+#ifdef HAVE_POLL_H
 #include <poll.h>
+#endif
+#ifdef HAVE_SYS_POLL_H
```
I don't see any `HAVE_SYS_POLL_H` in `pg_config.h.in`. Why are we doing this?

lithp replied:
Hmm, probably @mtuncer was trying to make this section match [the one here](https://github.com/citusdata/citus/pull/2011/files#diff-b95ac9c2bf85402fbb5eca5edcd94302R35), but that section looks like it was originally broken, we can probably just remove `HAVE_SYS_POLL_H` entirely

pykello commented on src/backend/distributed/connection/connection_management.c:
```diff
@@ -469,6 +485,31 @@ FinishConnectionEstablishment(MultiConnection *connection)
+				FD_SET(selectFileDescriptor, &writeFileDescriptorSet);
+			}
+
+			pollResult = (select) (selectFileDescriptor + 1, &readFileDescriptorSet,
```
why are the parenthesis around `select`?

lithp replied:
This might not be necessary since I think `FRONTEND` isn't defined while citus builds, but it was to make sure we don't hit [the `select` macro](https://github.com/postgres/postgres/blob/REL_10_STABLE/src/include/port/win32.h?utf8=%E2%9C%93#L374). We really just want to do our own thing and not think about whatever layer postgres tries to put between us and the syscall.

pykello commented on src/backend/distributed/connection/remote_commands.c:
```diff
@@ -994,14 +994,20 @@ BuildWaitEventSet(MultiConnection **allConnections, int totalConnectionCount,
 	int pendingConnectionCount = totalConnectionCount - pendingConnectionsStartIndex;
 	WaitEventSet *waitEventSet = NULL;
 	int connectionIndex = 0;
+	const int MAX_WAIT_EVENT_COUNT = 50;
```
This should probably be a function of `FD_SETSIZE` instead of constant 50 everywhere.

lithp replied:
Oh yeah, good point

lithp commented:
@pykello we're not using `WaitEventSet`s because we ran into multiple different issues with them, @mtuncer would be better able to tell you what he was seeing before we gave up and moved to `select()`, but once we switched to `select()` all of our problems went away.

mtuncer commented:
@pykello @lithp we were having connection establishment issues when we are opening many connections. When we use wait event set, for some reason connection state update from remote node did not made it back to the coordinator, we kept getting invalid connection state (or something similar) when we try to proceed with the connection establishment routine. We tried some workarounds but none of them worked, therefore we rolled back to our original (and somewhat old)  poll/select logic. 

@anarazel does not agree with this and says we should find a way to use wait event set which may also include fixing this issue  in postgres side. 

At this moment, we would go with using select (preferably for windows builds only).

lithp commented:
I think we should probably hold off on this change and do it the right way.

lithp commented:
https://github.com/citusdata/citus/pull/2011/commits/b3dd772217f3869070dc9b13b453326751b85ac8 seems to have broken the build on Windows.

lithp commented:
@pykello okay, I think we're going to merge this after all. Friendly ping asking for a review :)

sumedhpathak commented:
@lithp Let's hold on merging. I sent a mail about it.

pykello commented on src/backend/distributed/executor/multi_client_executor.c:
```diff
@@ -815,8 +814,19 @@ MultiClientCreateWaitInfo(int maxConnections)
+	/* we subtract 2 to make room for the WL_POSTMASTER_DEATH and WL_LATCH_SET events */
+	if (maxConnections > FD_SETSIZE - 2)
+	{
+		maxConnections = FD_SETSIZE - 2;
```
Do we need to do this when poll is available?

lithp replied:
Addressed in https://github.com/citusdata/citus/pull/2011/commits/114414d9f1f75e61476d2d13493ecb3668ab1505

pykello commented on src/backend/distributed/executor/multi_client_executor.c:
```diff
@@ -853,9 +876,17 @@ MultiClientRegisterWait(WaitInfo *waitInfo, TaskExecutionStatus executionStatus,
+	if (waitInfo->registeredWaiters >= waitInfo->maxWaiters)
+	{
+		return;
+	}
```
What would happen if a job needs more connections than `maxWaiters`? We either shouldn't allow more than `maxWaiters` connections by erroring out, or be sure that we can handle it.

lithp replied:
In that case we'll have worse-than-usual performance. We'll wait until one of the `maxWaiters` connections becomes ready. When a connection outside of that set becomes ready we won't react until one of the other connections triggers. Since we go down the list and try to use every connection whenever one of them has an event I don't expect that this would have a noticeable impact on performance.

pykello commented on src/backend/distributed/executor/multi_client_executor.c:
```diff
@@ -879,6 +911,26 @@ MultiClientRegisterWait(WaitInfo *waitInfo, TaskExecutionStatus executionStatus,
+
+	if (executionStatus == TASK_STATUS_SOCKET_READ)
+	{
+		FD_SET(connectionFileDescriptor, &(waitInfo->exceptionFileDescriptorSet));
```
Exception fd set isn't same as `POLLERR`. Based on the man pages, it is more like `POLLPRI`. So we don't need to set it here. Actually, we also don't need to set `POLLERR` for `poll`, since it is ignored in `events`.

lithp replied:
Addressed in https://github.com/citusdata/citus/pull/2011/commits/fd699f12534ebf78a84a0d6e77988d1a6961b0d9

lithp commented on src/backend/distributed/connection/connection_management.c:
```diff
@@ -469,6 +475,31 @@ FinishConnectionEstablishment(MultiConnection *connection)
+			pollResult = (select) (selectFileDescriptor + 1, &readFileDescriptorSet,
+								   &writeFileDescriptorSet, &exceptionFileDescriptorSet,
+								   &selectTimeout);
+#endif /* HAVE_POLL */
```
Addressed in https://github.com/citusdata/citus/pull/2011/commits/488e5d1fdb552ede7de3b1bd96494303e8884f04

lithp commented on src/backend/distributed/connection/connection_management.c:
```diff
@@ -10,7 +10,10 @@
 
+#ifdef HAVE_POLL_H
 #include <poll.h>
+#endif
```
I'm not entirely sure, somewhere inside "postgres.h" it's being included for us. I assume [from here](https://github.com/postgres/postgres/blob/9d4649ca49416111aee2c84b7e4441a0b7aa2fac/src/include/port/win32_port.h)

pykello commented on src/backend/distributed/executor/multi_client_executor.c:
```diff
@@ -832,14 +844,27 @@ MultiClientResetWaitInfo(WaitInfo *waitInfo)
 {
+#ifdef HAVE_POLL
 	pfree(waitInfo->pollfds);
+#else
```
what is the `#else` for?

pykello commented on src/backend/distributed/executor/multi_client_executor.c:
```diff
@@ -832,14 +844,27 @@ MultiClientResetWaitInfo(WaitInfo *waitInfo)
 	waitInfo->haveFailedWaiter = false;
+
+#ifdef HAVE_POLL
+#else
```
why not use `#ifndef` instead?
